### PR TITLE
13647 fixed wrong feature flag for checking admin fine-grained authz

### LINKED ISSUE #13647: Wrong feature flag used to decide if admin fine-grained-permissions should be enabled
### Describe the bug

In several places, the flag `Profile.Feature.AUTHORIZATION` is used to decide if fine-grained permissions should be used. However, the correct flag is `Profile.Feature.ADMIN_FINE_GRAINED_AUTHZ`.

### Version

19.0.1

### Expected behavior

`Profile.Feature.ADMIN_FINE_GRAINED_AUTHZ` is used to decide whether admin fine-grained permissions checks should be performed.

### Actual behavior

`Profile.Feature.AUTHORIZATION` is used to decide whether admin fine-grained permissions checks should be performed.

### How to Reproduce?

Check source code.

### Anything else?

Also see discussion in https://github.com/keycloak/keycloak/pull/12654

### PR DESCRIPTION
**Pull request by sschu**
Fixes https://github.com/keycloak/keycloak/issues/13647.

### CODE REVIEW

**Comment by sschu** [comment]
@pedroigor 

Here is the PR for fixing other wrong usages of the feature flag.

**Comment by pedroigor** [comment]
@sschu Thanks. Looks like some tests are failing because now the `authz` field in `MgmtPermissions` is null and no check exists if this reference is null.

**Comment by sschu** [comment]
I'll have a look.

**Comment by sschu** [comment]
@pedroigor I had to adapted some tests to also enable the profile feature correctly. I hope that does the trick.

**Comment by pedroigor** [comment]
@sschu Thanks again! Makes more sense for me too.

However, I'm afraid we still have some null checks to make like:

* https://github.com/keycloak/keycloak/blob/d3b43a9f59764b763bffc11ba1f087f929f433ef/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java#L126.
* https://github.com/keycloak/keycloak/blob/d3b43a9f59764b763bffc11ba1f087f929f433ef/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java#L524

The only reason why they don't fail is because of checks like https://github.com/keycloak/keycloak/blob/c3ea847b3eb2d9081279ed168c2a5704216f3139/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissionManagement.java#L36.

So the places still using the `authz` even though `null` are only called if permissions are enabled for a specific area.

**Comment by sschu** [comment]
@pedroigor 
I also saw the the code for the `XyzPermissions` classes is not very safe and relies on being used in a safe way from external. I did not like that so I added another commit for easier review adding checks. I am somewhat confident I caught all cases. 

The general idea is to only check for the profile feature in one place in 'MgmtPermissions' and use `authz!=null` as an equivalent internal check down the road (not sure on this one, maybe checking for the feature directly is more explicit - but its also more code to read). In general `ADMIN_FINE_GRAINED_AUTHZ` enabled should be equivalent to `authz!=null`. Also, 'resourceServer!=null' means 'authz!=null' so it is safe to make calls to it. I am not 100% sure the reverse is also true so maybe there are cases where `resourceServer==null` and `authz!=null`. But most checks just check  `resourceServer!=null` and can safely use `authz` afterwards.

I also added some source code comments/questions, feel free to respond to them. If you like the changes I would finalise the second commit and finally squash it before merging.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java [hunk 1]
```diff
@@ -292,6 +293,7 @@ public Resource resource(ClientModel client) {
     @Override
     public Map<String, String> getPermissions(ClientModel client) {
         initialize(client);
+        //does it make sense to return the scopes if fine-grained permissions are disabled?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:296]: I don't think it does. The calling code does not call this method if the feature is disabled. See https://github.com/keycloak/keycloak/blob/2b996b12a180c426bbefe5cb44b6fce1af982840/services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java#L660.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/GroupPermissions.java [hunk 2]
```diff
@@ -222,6 +222,7 @@ public Resource resource(GroupModel group) {
     @Override
     public Map<String, String> getPermissions(GroupModel group) {
         initialize(group);
+        //does it make sense to return the scopes if fine-grained permissions are disabled?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/GroupPermissions.java:225]: Same comment and code pattern from the previous comments on clients.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/IdentityProviderPermissions.java [hunk 3]
```diff
@@ -140,6 +141,7 @@ public Resource resource(IdentityProviderModel idp) {
     @Override
     public Map<String, String> getPermissions(IdentityProviderModel idp) {
         initialize(idp);
+        //does it make sense to fill the scopes when fine-grained permissions are disabled?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/IdentityProviderPermissions.java:144]: Same comment and code pattern from the previous comments on clients.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/MgmtPermissions.java [hunk 4]
```diff
@@ -203,6 +203,7 @@ public RealmModel adminsRealm() {
 
     @Override
     public RolePermissions roles() {
+        //Is there any reason why role permissions are always created new while the other permissions are kept?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/MgmtPermissions.java:206]: I don't think so. We can do the same as others and see how it goes.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java [hunk 5]
```diff
@@ -91,6 +91,7 @@ private void disablePermissions(RoleModel role) {
     @Override
     public Map<String, String> getPermissions(RoleModel role) {
         initialize(role);
+        //does it make sense to initialize the scopes when fine-grained permissions are disabled?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java:94]: Same comment and code pattern from the previous comments on clients.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java [hunk 6]
```diff
@@ -602,6 +604,7 @@ private String getMapCompositePermissionName(RoleModel role) {
         return MAP_ROLE_COMPOSITE_SCOPE + ".permission." + role.getId();
     }
 
+    //What is this? Did somebody roll over the keyboard?;) Apparently not used
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java:607]: Definitely :) Let's remove it.

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/UserPermissions.java [hunk 7]
```diff
@@ -145,6 +145,7 @@ private void initialize() {
     @Override
     public Map<String, String> getPermissions() {
         initialize();
+        //does it make sense to initialize the scopes when fine-grained permissions are disabled?
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/UserPermissions.java:148]: Same comment and code pattern from the previous comments on clients.

**Comment by pedroigor** [comment]
@sschu I'm really pleased to collaborate with you on these changes and your willingness to improve this area. Thanks a lot for your time and effort.

Despite the answers to some of your comments, the proposed changes LGTM and they should help us to create a baseline to perform more aggressive refactoring in the future.

**Comment by sschu** [comment]
@pedroigor Thanks for the roses! ;) I am also happy to contribute. I implemented your suggestions. Should I then squash the commits if everything is fine or can it be automatically squash-merged?

**Comment by pedroigor** [comment]
I can squash before merging.

**Comment by sschu** [comment]
Hmm, the failing test ran fine locally.

**Comment by sschu** [comment]
@pedroigor The test failure is most likely not related to my changes.

**Review by pedroigor — APPROVED** [review]

**Pull request merged**